pkg/server/server.go

```diff
@@ -60,7 +63,23 @@ func (s *Server) ListRules(w http.ResponseWriter, r *http.Request, tenant string
 func (s *Server) SetRules(w http.ResponseWriter, r *http.Request, tenant string) {
 	logger := log.With(s.logger, "handler", "setrules", "tenant", tenant)
 
-	err := s.bucket.Upload(r.Context(), getRulesFilePath(tenant), r.Body)
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "reading request body", http.StatusInternalServerError)
+		level.Warn(logger).Log("msg", "reading request body", "err", err)
+
+		return
+	}
+	defer r.Body.Close()
+
+	if _, errs := rulefmt.Parse(data); errs != nil {
```

Comment:
This is confusing, I had to look at the documentation to find out it also does validation on top of parsing. But probably nothing we can change here.